docs/reactivity.md

```diff
@@ -0,0 +1,410 @@
+
+```js run=false
+const name = Generators.observe((notify) => {
+  const inputted = () => notify(nameInput.value);
+  inputted();
+  nameInput.addEventListener("input", inputted);
+  return () => nameInput.removeEventListener("input", inputted);
+});
+```
+
+As another example, here is using `Generators.observe` to expose the current pointer coordinates:
+
+```js echo
+const pointer = Generators.observe((change) => {
+  const pointermoved = (event) => change([event.clientX, event.clientY]);
+  addEventListener("pointermove", pointermoved);
+  change([0, 0]);
+  return () => removeEventListener("pointermove", pointermoved);
+});
+```
+
+```js echo
+pointer.map(Math.round) // try moving your mouse
+```
+
+And here’s a generator `j` that increments once a second, defined directly by an immediately-invoked async generator function.
+
+```js echo
+const j = (async function* () {
+  for (let j = 0; true; ++j) {
+    yield j;
+    await new Promise((resolve) => setTimeout(resolve, 1000));
+  }
+})();
+```
+
+```js echo
+j
+```
+
+If a generator does not explicitly `await`, it will yield once every [animation frame](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame), typically 60 times per second. Generators also automatically pause when the page is put in a background tab.
+
+```js echo
+const i = (function* () {
+  for (let i = 0; true; ++i) {
+    yield i;
+  }
+})();
+```
+
+```js echo
+i
+```
+
+As you might imagine, you can use such a generator to drive an animation. A generator is typically easier than a `requestAnimationFrame` loop because the animation is declarative — the code runs automatically whenever `i` changes — and because you don’t have to handle [invalidation](#invalidation) to terminate the loop.
+
+```svg echo
+<svg width="640" height="32">
+  <rect fill="#4269d0" width="32" height="32" x=${(i % (640 + 32)) - 32}></rect>
+</svg>
+```
+
+You can also use a generator to stream live data. Here is a WebSocket that listens for the current price of Bitcoin, keeping the last minute of data in memory.
+
+```js echo
+const socket = new WebSocket("wss://ws.eodhistoricaldata.com/ws/crypto?api_token=demo");
+invalidation.then(() => socket.close());
+socket.addEventListener("open", () => socket.send(JSON.stringify({action: "subscribe", symbols: "BTC-USD"})));
+const messages = Generators.observe((change) => {
+  const messages = [];
+  const duration = messages.duration = 60_000;
+  const messaged = (event) => {
+    const m = JSON.parse(event.data);
+    const t = m.t;
+    if (t == null) return;
+    while ((t - messages[0]?.t) > duration) messages.shift();
+    messages.push(m);
+    change(messages);
+  };
+  socket.addEventListener("message", messaged);
+  return () => socket.removeEventListener("message", messaged);
+});
+```
+
+```js echo
```

Comment:
This is example is not working correctly for me (2024/04/11, on Firefox).

Reply:
This works for me on Firefox, but it depends on an external API (because the point is to demonstrate real-time streaming data over a websocket), so it’s possible either the API was down, or that it’s not accessible on your network e.g. due to firewall restrictions.